packages/pluggableWidgets/datagrid-web/src/utils/settings.ts

```diff
@@ -0,0 +1,116 @@
+import { Filters, IdType, SortingRule } from "react-table";
+import { Dispatch, SetStateAction, useEffect, useMemo, useState } from "react";
+import { EditableValue, ValueStatus } from "mendix";
+import deepEqual from "deep-equal";
+import { ColumnWidth, TableColumn } from "../components/Table";
+
+declare type Option<T> = T | undefined;
+
+interface Settings {
+    columnOrder: Array<IdType<object>>;
+    hiddenColumns: Array<IdType<object>>;
+    sortBy: Array<SortingRule<object>>;
+    filters: Filters<object>;
+    widths: ColumnWidth;
+}
+
+interface PersistedSettings {
+    column: string;
+    sort: boolean;
+    sortMethod: "asc" | "desc";
+    filter: string;
+    hidden: boolean;
+    order: number;
+    width: number | undefined;
+}
+
+export function createSettings(
+    { columnOrder, hiddenColumns, sortBy, filters, widths }: Settings,
+    columns: Array<{ header: string; id: string }>
+): PersistedSettings[] {
+    return columns.map(column => ({
+        column: column.header,
+        sort: !!sortBy.find(s => s.id === column.id),
+        sortMethod: sortBy.find(s => s.id === column.id)?.desc ? "desc" : "asc",
+        filter: filters.find(f => f.id === column.id)?.value ?? "",
+        hidden: !!hiddenColumns.find(h => h === column.id),
+        order: columnOrder.findIndex(o => o === column.id),
+        width: widths[column.id]
+    }));
+}
+
+export function useSettings(
+    settings: Option<EditableValue<string>>,
+    columns: TableColumn[],
+    columnOrder: Array<IdType<object>>,
+    setColumnOrder: Dispatch<SetStateAction<Array<IdType<object>>>>,
+    hiddenColumns: Array<IdType<object>>,
+    setHiddenColumns: Dispatch<SetStateAction<Array<IdType<object>>>>,
+    sortBy: Array<SortingRule<object>>,
+    setSortBy: Dispatch<SetStateAction<Array<SortingRule<object>>>>,
+    filters: Filters<object>,
+    setFilters: Dispatch<SetStateAction<Filters<object>>>,
+    widths: ColumnWidth,
+    setWidths: Dispatch<SetStateAction<ColumnWidth>>
+): void {
+    const [loaded, setLoaded] = useState(false);
+
+    const filteredColumns = useMemo(
+        () =>
+            columns.map((c, index) => ({
+                header: typeof c.header === "object" ? c.header.value : c.header,
+                id: index.toString(),
+                hidable: c.hidable
+            })) as Array<{ header: string; id: string; hidable: string }>,
+        [columns]
+    );
+
+    useEffect(() => {
+        if (settings && settings.status !== ValueStatus.Loading && settings.value && !loaded) {
```

Comment:
If is switches back to loading the whole table will be reloaded as i am passing settings as an attribute of the hook

Reply:
Yes, I see. But it will wait again till the data is loaded. It could show in the meantime the previous value to prevent flickering. See the part about status: https://docs.mendix.com/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets#dynamic-value

Maybe this is something to think about in general for the data grid widget in a separate story.

Reply:
this is just loaded once :)